### ISSUE TITLE
rtorrent starts up ok then eventually goes to 100% cpu usage and spits out failed to connect and connection to all trackers halt.

### ISSUE THREAD

**Issue by evilbulgarian** [rakshasa/rtorrent#1208]
I have about 2100 torrents, all new torrents seem to not even start when added via flexget. I am attaching my config and emerge --info. I have tried recompiling curl with nss only support same issue.

[gypsyops_emerge.info.txt](https://github.com/rakshasa/rtorrent/files/10610048/gypsyops_emerge.info.txt)
[rtorrent.rc.txt](https://github.com/rakshasa/rtorrent/files/10610049/rtorrent.rc.txt)

```
1675560638 I 55565E134F5D55B1E4B6372D709E61046211D0E9->tracker_list: failed to connect to tracker ...
1675561079 I handshake_manager->108.50.139.230: Received error: message:7 network error.
1675561151 I handshake_manager->75.82.54.146: Received error: message:7 network unreachable.
```

**Comment by evilbulgarian** [comment]
```
❯ curl --version
curl 7.87.0 (x86_64-pc-linux-gnu) libcurl/7.87.0 NSS/3.79.2 zlib/1.2.13 zstd/1.5.2 c-ares/1.18.1 nghttp2/1.51.0
Release-Date: 2022-12-21
Protocols: dict file ftp ftps http https imap imaps mqtt pop3 pop3s rtsp smtp smtps tftp
Features: AsynchDNS HTTP2 HTTPS-proxy IPv6 Largefile libz NTLM SSL threadsafe UnixSockets zstd
```

**Comment by kannibalox** [comment]
Did rTorrent work for you built with earlier versions of curl, or has it been behaving like this all the time?

**Comment by evilbulgarian** [comment]
I have 2 other servers and they seem to be ok with the curl compiled with openssl. Going to set up debug logging on them and check for same errors. It something that happens over time it seems.

**Comment by JohnFlowerful** [comment]
It was fine with at least curl 7.84.0. This version (amongst others) has been [removed from package managers](https://curl.se/docs/security.html). Maybe a curl regression similar to #580  

New torrents start working eventually for me though.

**Comment by evilbulgarian** [comment]
> It was fine with at least curl 7.84.0. This version (amongst others) has been [removed from package managers](https://curl.se/docs/security.html). Maybe a curl regression similar to #580
> 
> New torrents start working eventually for me though.
> 
> Edit: I installed [net-misc/curl-7.86.0-r3](https://gitweb.gentoo.org/repo/gentoo.git/tree/net-misc/curl/curl-7.86.0-r3.ebuild?id=00add340e59844d43b0704b8e686c47aacea7afa) and can confirm that this resolved the issue for me.

I downgraded to same version still startup is fine then goes to 100% cpu with same network error messages :( thanks for suggestion tho.

**Comment by evilbulgarian** [comment]
OK, I think that the reverted curl actually is making my load avg a lot better its not constantly at 100% cpu only once in a while. Do you think we should submit a bug in gentoo as well @JohnFlowerful?

**Comment by JohnFlowerful** [comment]
My issue is fixed with now in-tree [net-misc/curl-7.88.0-r1](https://packages.gentoo.org/packages/net-misc/curl). I'm not sure of the cause, but curl's issue tracker had a few people reporting inconsistencies between 7.86 and 7.87. I'd be guessing, but https://github.com/curl/curl/pull/10160 looks relevant. 

I don't think it's a Gentoo issue. If it's still an issue with the new version, you're better off reporting directly to curl.

**Comment by stickz** [comment]
Does this pull request help your issue? Your load average will be significantly lower. It will just take a little bit more time for trackers to update initially. https://github.com/rakshasa/rtorrent/pull/1131

**Comment by tholin** [comment]
I might be experiencing the same problem. Rtorrent sometimes get stuck using 100% cpu for several seconds or even minutes at a time. I looked into the problem and it turns out to be a regression in Gentoo's net-misc/curl.

The ebuild for curl 7.84.0 and before had a default disabled "threads" useflag but more recent version removed the useflag and defaults to --enable-pthreads and --enable-threaded-resolver. This is the change that cause rtorrent to use 100% cpu. If I build curl-7.87.0-r2 with USE="-adns" and modify the ebuild to disable threaded-resolver the problem goes away.

Rtorrent use 100% cpu because it's stuck calling the epoll system call in a busy loop. Without a multithreaded resolver I assume rtorrent would just block during name resolves. With asynchronous resolves rtorrent keeps executing and goes wild.

Both evilbulgarian and JohnFlowerful seems to be Gentoo users so I'm guessing it's the same problem. All Gentoo versions of curl after 7.84.0 after broken for me since they all force --enable-threaded-resolver. The problem only triggers when DNS resolves are a bit flakey and have a risk to timeout.

**Comment by anzz1** [comment]
@tholin 
Is the latest rtorrent release still built with newer worse curl and the threaded resolver? This seems to be exactly the problem I'm having too as the stalling and CPU going to 100% seems to happen when its going through the tracker list and resolving hosts.
It would also explain why this problem happens now with the newer rtorrent from the debian 12 packages when older version on debian 7 worked just fine and had very low cpu usage all the time.

I don't have that many torrents just ~10 and it completely freezes up on startup and also periodically and now it all makes sense if its the DNS resolver busylooping for some tracker that does not have a DNS record as in the domain no longer exist.

Unfortunately I don't exactly remember what version of rtorrent i installed from the debian 7 package manager it probably wasn't the latest one available there either and I never updated it but it ran extremely solid for 10 years straight I wish I'd written down the version number so I could go back on it and never change it

**Comment by rakshasa** [comment]
Closing this as it seems to be a libcurl issue.

**Comment by anzz1** [comment]
Well it seems that from the rtorrent side you should have configured libcurl to not use the threaded resolver? 

> Rtorrent use 100% cpu because it's stuck calling the epoll system call in a busy loop. Without a multithreaded resolver I assume rtorrent would just block during name resolves. With asynchronous resolves rtorrent keeps executing and goes wild.

So rtorrent makes a spinning thread instead of a blocking one because of misconfiguration with libcurl?

**Comment by rakshasa** [comment]
Rtorrent does what libcurl tells it to do, and if libcurl doesn't tell us to remove an fd from polling then that's a libcurl bug.

Doesn't matter how libcurl was configured, it's a very bad bug on their side.

**Comment by anzz1** [comment]
So rtorrent can go to an older libcurl version that doesn't have this bug? Or which rtorrent version used older libcurl so I can go to older rtorrent version to fix it on my end? Like said I used to have some old rtorrent version taken from debian 7 repo I think 12 years ago and it had a version that was rock solid and ran with minimal cpu needs and never crashed or had any bugs whatsoever for 10+ years straight. Wonderful piece of software that was.

**Comment by rakshasa** [comment]
Just compile with an older libcurl version, however this isn't the place for basic user-support so ask google.

**Comment by anzz1** [comment]
Yeah alright for reference for others i'm quite sure the version I had that worked extremely well was rtorrent 0.9.2 with libtorrent 0.13.2 . and the libcurl is just a guess but that'd be 7.38.0 configured with --disable-thread-resolver . 

https://github.com/openssl/openssl/releases/download/OpenSSL_1_0_2u/openssl-1.0.2u.tar.gz
https://curl.se/download/curl-7.38.0.tar.gz
https://archive.debian.org/debian/pool/main/r/rtorrent/rtorrent_0.9.2.orig.tar.gz
http://archive.debian.org/debian/pool/main/libt/libtorrent/libtorrent_0.13.2.orig.tar.gz

That's most likely the setup that is stable and high performance but its going to be bitch and a half to compile as modern GCC breaks everything.

**Comment by tholin** [comment]
Back when I encountered this problem I noticed that I could work around it by setting the envvar "RTORRENT_POLL=select" and I have been using that workaround ever since.

When reviewing the rtorrent source code just now I can see that the RTORRENT_POLL option was removed in rtorrent-0.15.4 (commit 4c63cee7a26d685b2673d7978d96b4a46c47d67b), and I'm using version 0.15.7. Despite this I am not experiencing any 100% CPU usage issues.

Gentoo's stable curl-8.19.0.ebuild contains the following section:

    # Since 8.12.0 adns/c-ares and the threaded resolver are mutually exclusive
    # This is in support of some work to enable `httpsrr` to use adns and the rest
    # of curl to use the threaded resolver; for us `httpsrr` is conditional on adns.
    if use adns; then
            myconf+=(
                    --disable-threaded-resolver
            )
    else
            myconf+=(
                    --enable-threaded-resolver
            )
    fi

The adns useflag is enabled by default so it seems that Gentoo is once again disabling the threaded resolver by default. I suspect that is why I am no longer experiencing the problem.

For reference here is how my curl is configured:

    $ curl --version
    curl 8.19.0 (x86_64-pc-linux-gnu) libcurl/8.19.0 OpenSSL/3.5.6 zlib/1.3.2 c-ares/1.34.6 libpsl/0.21.5 nghttp2/1.68.0 ngtcp2/1.22.0 nghttp3/1.9.0
    Release-Date: 2026-03-11
    Protocols: dict file ftp ftps http https imap imaps ipfs ipns mqtt mqtts pop3 pop3s rtsp smtp smtps tftp ws wss
    Features: alt-svc AsynchDNS HSTS HTTP2 HTTP3 HTTPS-proxy HTTPSRR IPv6 Largefile libz NTLM PSL SSL threadsafe TLS-SRP UnixSockets

**Comment by anzz1** [comment]
@tholin 
Thank you a ton, GCC 12 really didn't want to build rtorrent 0.9.2, but I saw that in the debian repo there is 0.9.8 for Debian 12 and that version still has the `RTORRENT_POLL` envvar, which seems to have completely eliminated the problem and instead of jumping to 100% now rtorrent uses 1% cpu at most. It's confusing though considering that epoll/poll/select should all be event-driven and none of those should be the cause of any spinning busy loops unless something is done wrong like the wakeup period set to zero and looped. In fact epoll should be the fastest and lightest weight version of them all. That being said historically there have been all sorts of bugs and quirks with the linux polling systems which is why it used to be a norm that its configurable which polling method to use and rtorrent was no different in this regard. Shame to see that option go, especially as it fixed the issue.

Anyway `RTORRENT_POLL=select` on rtorrent 0.9.8/libtorrent 0.13.8 fixes the high cpu usage problem and I don't have any other problems so far except the ones I always had with the newer versions. The intermittently hanging UI stays and SIGTERM kill still takes an extraordinarily long time (30sec-1min), but CTRL+Q worked for closing the app for the first time after setting the env var. Before it used to just close the torrents but never close the app.

<img width="1063" height="355" alt="Image" src="https://github.com/user-attachments/assets/68a31817-22c9-4ccf-a347-e8ce244862a7" />

At 20:20 I replaced the rtorrent with 0.9.8 one and `RTORRENT_POLL=select` and spikes are gone. With 8 logical cores that jump to 12.5% means spikes to 100% lock of a single core. All gone now :) Thanks again so much tholin love that the solution was ultimately so simple, just an env variable.

**Comment by tholin** [comment]
There have been some changes upstream in curl with two patches in 8.21.0 with fixes for this problem.

https://github.com/curl/curl/commit/2a2104f3cff44bb28bb570a093be52bbeeed8f23
https://github.com/curl/curl/commit/5e4e62962c0f8536c612a5248e2ac4eb7d67ae66

Great, but not really because the latter patch 5e4e62962c0f8536c612a5248e2ac4eb7d67ae66 actually introduces busy poll looping for me. I had to do a bisect 8.20.0 -> 8.21.0 to find the culprit. Perhaps the patches work for the threaded-resolver case but break when the threaded-resolver is not used?

The current git version of curl is unaffected so I did another bisect to see what changed. The patch https://github.com/curl/curl/commit/820c014578b56b9fa4d2c472e90b5060198dd404 fixes the busy looping introduced in 8.21.0 but it is not yet part of any release.

TL;DR curl 8.21.0 is broken. There is a fix but it is not included in any release yet. Avoid 8.21.0 for the time being.

**Comment by anzz1** [comment]
Good work @tholin on your thorough research. That being said, I don't really see any reason updating from rtorrent 0.9.8/libtorrent 0.13.8 as that seems to be quite stable with the `RTORRENT_POLL=select` option.

Don't fix what ain't broken usually works well in computing, if just there weren't so many pesky people going "but iiitssss *deprecated* you shouldn't use that, you should use this shiny new thing instead, which is often worse than the old thing but hey at least this one is *new* and not *deprecated*".
Linux is especially bad on this, windows for its many faults breaking backwards compatibility isn't one of them. On Linux, I really couldn't compile (nor run a precompiled one) the rtorrent 0.9.2 which was my original goal even if I wanted to, its like GCC goes out of its way to break compiling older codebases as much as possible. And the *nix dependency hell, stuff of nightmares. On Windows I can run utorrent 2.0.4 build 22967, the perfect gold standard client, released 16 years ago with zero issues, as if it was released today. But Linux, uuuugghhh, snowball's chance in hell for anything like that. And let's not even go into things like deciding to rewrite core tools that have been hardened and perfected for 40 years with something *new* written in *Rust*.
